openfl-tutorials/interactive_api/PyTorch_MVTec_PatchSVDD/workspace/PatchSVDD_with_Director.ipynb

```diff
@@ -0,0 +1,1211 @@
+    "    else:\n",
+    "        device = 'cuda'\n",
+    "    \n",
+    "    function_defined_in_notebook(some_parameter)\n",
+    "    \n",
+    "    train_loader = tqdm.tqdm(train_loader, desc=\"train\")\n",
+    "    model.modelA = enc\n",
+    "    model.modelB = cls_64\n",
+    "    model.modelC = cls_32\n",
+    "    model_list = [enc, cls_64, cls_32]\n",
+    "    losses = []\n",
+    "    \n",
+    "    for module in model_list:\n",
+    "        module.train()\n",
+    "\n",
+    "    for d in train_loader:\n",
+    "        d = to_device(d, 'cuda', non_blocking=True)\n",
+    "        optimizer.zero_grad()\n",
+    "        loss_pos_64 = PositionClassifier.infer(cls_64, enc, d['pos_64'])\n",
+    "        loss_pos_32 = PositionClassifier.infer(cls_32, enc.enc, d['pos_32'])\n",
+    "        loss_svdd_64 = SVDD_Dataset.infer(enc, d['svdd_64'])\n",
+    "        loss_svdd_32 = SVDD_Dataset.infer(enc.enc, d['svdd_32'])\n",
+    "\n",
+    "        loss = loss_pos_64 + loss_pos_32 + float(args['lambda_value']) * (loss_svdd_64 + loss_svdd_32)\n",
+    "        loss.backward()\n",
+    "        optimizer.step()\n",
+    "        losses.append(loss.detach().cpu().clone().numpy())\n",
+    "    return {'train_loss': np.mean(losses),}\n",
+    "    \n",
+    "\n",
+    "@TI.register_fl_task(model='model', data_loader='val_loader', device='device')     \n",
+    "def validate(model, val_loader, device):\n",
+    "    if not torch.cuda.is_available():\n",
+    "        device = 'cpu'\n",
+    "    else:\n",
+    "        device = 'cuda'\n",
+    "    \n",
+    "    model.modelA = enc\n",
+    "    enc.eval()\n",
+    "    enc.to(device)\n",
+    "    \n",
+    "    x_tr, x_te, masks, labels, mean = val_loader\n",
+    "\n",
+    "    embs64_tr = infer_(x_tr, enc, K=64, S=16)\n",
+    "    embs64_te = infer_(x_te, enc, K=64, S=16)\n",
+    "    embs32_tr = infer_(x_tr, enc.enc, K=32, S=4)\n",
+    "    embs32_te = infer_(x_te, enc.enc, K=32, S=4)\n",
+    "\n",
+    "    embs64 = embs64_tr, embs64_te\n",
+    "    embs32 = embs32_tr, embs32_te\n",
+    "\n",
+    "    results = eval_embeddings_nn_multik(args['obj'], embs64, embs32, masks, labels)\n",
+    "    \n",
+    "    maps = results['maps_mult']\n",
+    "    obj = args['obj']\n",
+    "\n",
+    "    print(\"| K64 | Det: {:.3f} Seg:{:.3f} BA: {:.3f}\".format(results['det_64'],results['seg_64'],results['bal_acc_64']))\n",
+    "    print(\"| K32 | Det: {:.3f} Seg:{:.3f} BA: {:.3f}\".format(results['det_32'],results['seg_32'],results['bal_acc_32']))\n",
+    "    print(\"| sum | Det: {:.3f} Seg:{:.3f}\".format(results['det_sum'],results['seg_sum']))\n",
+    "    print(\"| mult | Det: {:.3f} Seg:{:.3f}\".format(results['det_mult'],results['seg_mult']))\n",
+    "\n",
+    "    return {'detection_score': results['det_sum'], 'segmentation_score': results['seg_sum'], 'balanced_accuracy_score': results['bal_acc_64']}\n",
+    "\n",
+    "\n",
+    "def infer_(x, enc, K, S):\n",
```

Comment:
Can we move `infer_()` function to a separate cell, since it looks like it's not an FL task (it's not decorated with `@TI.register_fl_task()`)